ZcashFoundation/zebra
# Fix minimum difficulty bugs

## Motivation

There are a bunch of bugs in Zebra's difficulty implementation, and the Zcash difficulty specifications also have some bugs.

## Solution

New changes:
- [x] Revise the testnet minimum difficulty code based on `zcashd`'s implementation zcash/zips#416
  - ZIPs 205 and 208 are wrong 
  - the min-difficulty rules change difficulty adjustment, not the difficulty filter
  - all blocks pass the default difficulty filter
- [x] Turn `expect("minimum difficulty activation is checked earlier in the test")` into an `Err` if the rule is not active (and so the function returns `None`)

Tests for spec errors:
  - [x] Test that testnet min-difficulty rules follow `zcashd` https://github.com/zcash/zips/issues/416
  - [x] Test implicit conversion of the minimum difficulty threshold to the compact representation #1277
    - covered by the genesis difficulty equality test

Reviewed changes:
- Refactor block target spacing into NetworkUpgrade methods
- Make ExpandedDifficulty Debug byte order match Hash
  - Bug fix on PR #1171
  - Without this change, test failures are really confusing
- Round-trip the PoWLimit through the compact representation
  - the Zcash specification is ambiguous, see #1277
- Add minimum difficulty testnet block vectors
- Add tests for testnet minimum difficulty blocks
  - the Zcash specification is ambiguous, see #1276
  - find a pre-Blossom and post-Blossom minimum difficulty block, and add them to `zebra_test::vectors::block`, along with their previous blocks
  - make sure the `BlockVerifier` difficulty tests pass on these new blocks

The code in this pull request has:
  - [x] Documentation Comments
  - [x] Unit Tests and Property Tests

## Review

@yaahc has reviewed the Difficulty Contextual Validation RFC, which describes these checks.

## Related Issues

Difficulty Contextual Validation RFC PR #1246 
Difficulty tracking issue #802

Fixes bugs in:
- Previous difficulty check PR #809 
- Reverse the displayed endianness of hashes #1171
- Off-by-one error in testnet minimum difficulty block gap specification #1276 (Zcash specification)
- Implicit conversion of the minimum difficulty threshold to the compact representation #1277 (Zcash specification)
- Testnet min-difficulty rules are incorrectly specified https://github.com/zcash/zips/issues/416

## Follow Up Work

- ~Move the difficulty filter check into the state service~ - obsoleted by zcash/zips#416
- allow minimum difficulty testnet blocks
- Change the difficulty threshold check into a method on `Block` or `ExpandedDifficulty`
  - it's easy to get `<=` confused with `<`

This work is listed in the difficulty tracking issue #802
